falcon/media/multipart.py

```diff
@@ -0,0 +1,402 @@
+        #   (yet) for parsing the prologue. The CRLF will be prepended later to
+        #   construct the inter-part delimiter as per the RFC 7578, section 4.1
+        #   (see the note below).
+        self._dash_boundary = b'--' + boundary
+        self._parse_options = parse_options
+
+    def __iter__(self):
+        delimiter = self._dash_boundary
+        stream = self._stream
+        max_headers_size = self._parse_options.max_body_part_headers_size
+        remaining_part_count = self._parse_options.max_body_part_count
+
+        while True:
+            # NOTE(vytas): Either exhaust the unused stream part, or skip
+            #   the prologue.
+            stream.pipe_until(delimiter)
+            stream.read(len(delimiter))
+
+            if not delimiter.startswith(CRLF):
+                # NOTE(vytas): RFC 7578, section 4.1.
+                #   As with other multipart types, the parts are delimited with
+                #   a boundary delimiter, constructed using CRLF, "--", and the
+                #   value of the "boundary" parameter.
+                delimiter = CRLF + delimiter
+
+            separator = stream.read_until(CRLF, 2, MultipartParseError)
+            if separator == b'--':
+                if stream.peek(2) != CRLF:
+                    raise MultipartParseError('unexpected form epilogue')
+                break
+            elif separator:
+                raise MultipartParseError('unexpected form structure')
+
+            headers = {}
+            headers_block = stream.read_until(_CRLF_CRLF, max_headers_size,
+                                              MultipartParseError)
+            stream.read(4)
+
+            for line in headers_block.split(CRLF):
+                name, sep, value = line.partition(b': ')
+                if sep:
+                    name = name.lower()
+
+                    # NOTE(vytas): RFC 7578, section 4.5.
+                    #   This use is deprecated for use in contexts that support
+                    #   binary data such as HTTP. Senders SHOULD NOT generate
+                    #   any parts with a Content-Transfer-Encoding header
+                    #   field.
+                    #
+                    #   Currently, no deployed implementations that send such
+                    #   bodies have been discovered.
+                    if name == b'content-transfer-encoding':
+                        raise MultipartParseError(
+                            'the deprecated Content-Transfer-Encoding header '
+                            'field is unsupported')
+                    # NOTE(vytas): RFC 7578, section 4.8.
+                    #   Other header fields MUST NOT be included and MUST be
+                    #   ignored.
+                    elif name in _ALLOWED_CONTENT_HEADERS:
+                        headers[name] = value
+
+            remaining_part_count -= 1
+            if remaining_part_count == 0:
```

Comment:
This looks like an off-by-one error. I.e., if `max_body_part_count` is `1`, `remaining_part_count` is initialized to `1`. Then, after consuming the first part, `remaining_part_count` becomes `0` and `MultipartParseError` is raised even though we have only consumed more than `max_body_part_count` yet. I haven't checked the tests yet, but there should be one for this boundary condition.

Reply:
Touché; there is a test for this boundary, but it is off-by-one from this boundary condition :man_facepalming: